File: Stack/Easy/infix_to_postfix.cpp
// https://practice.geeksforgeeks.org/problems/infix-to-postfix-1587115620/1#
// { Driver Code Starts
// C++ implementation to convert infix expression to postfix
#include <bits/stdc++.h>
using namespace std;


 // } Driver Code Ends
class Solution {
  public:
    // Function to convert an infix expression to a postfix expression.
    int prec(char c)
    {
       if(c=='^'){return 3;}
       else if(c=='/'||c=='*'){return 2;}
       else if(c=='+'||c=='-'){return 1;}
       else{return -1;}
    }
   
   bool isOperand(char ch)
   {
       if(ch >='a' && ch <='z' || ch >='A' && ch <='Z')
            return true;
        return false;
   }
   
   string infixToPostfix(string s) {
       // Your code here
       stack<char> st;
       string ans;
       for(int i=0;i<s.length();i++){
           if(isOperand(s[i])){
               ans+=s[i]; //add the character by concatenation
           }
           else if(s[i]=='('){
               st.push(s[i]);
           }
           else if(s[i]==')'){
               while(st.top()!='('){
                   ans+=st.top();
                   st.pop();
               }
               st.pop();
           }
           else{
               while(!st.empty() && prec(s[i]) <= prec(st.top())){
                   ans += st.top();
                   st.pop();
               }
               st.push(s[i]);
           }
       }
       while(!st.empty()){
           ans+=st.top();
           st.pop();
       }
       return ans;
    }
};


// { Driver Code Starts.
// Driver program to test above functions
int main() {
    int t;
    cin >> t;
    cin.ignore(INT_MAX, '\n');
    while (t--) {
        string exp;
        cin >> exp;
        Solution ob;
        cout << ob.infixToPostfix(exp) << endl;
    }
    return 0;
}
  // } Driver Code Ends

File: Stack/Easy/infix_to_postfix.py
# https://practice.geeksforgeeks.org/problems/infix-to-postfix-1587115620/1#
#User function Template for python3


class Solution:
    
    #Function to convert an infix expression to a postfix expression.
    def getPrecedence(self, ch)->int:
        if(ch == '^'):
            return 3
        if(ch == '/' or ch == '*' ):
            return 2
        if(ch == '+' or ch == '-'):
            return 1
        return -1
        
    def isOperand(self, ch):
        return ch.isalpha()
        
    def InfixtoPostfix(self, exp):
        ans = ""
        st = []
        for ch in exp:
            if(self.isOperand(ch)):
                ans += ch
                
            elif(ch == '('):
                st.append(ch)
                
            elif(ch == ')'):
                while(st[-1] != '('): 
                    top = st.pop()
                    ans += top
                st.pop()
                
            else:
                while(len(st) > 0 and (self.getPrecedence(st[-1]) >= self.getPrecedence(ch))):
                    top = st.pop()
                    ans += top
                    
                st.append(ch)
                
        while(len(st) > 0):
            top = st.pop()
            ans += top
            
        return ans;
#{ 
#  Driver Code Starts
#Initial Template for Python 3

import atexit
import io
import sys

# This code is contributed by Nikhil Kumar Singh(nickzuck_007)


_INPUT_LINES = sys.stdin.read().splitlines()
input = iter(_INPUT_LINES).__next__
_OUTPUT_BUFFER = io.StringIO()
sys.stdout = _OUTPUT_BUFFER

@atexit.register

def write():
    sys.__stdout__.write(_OUTPUT_BUFFER.getvalue())


if __name__ == '__main__':
    test_cases = int(input())
    for cases in range(test_cases) :
        exp = str(input())
        ob=Solution()
        print(ob.InfixtoPostfix(exp))
# } Driver Code Ends